File: backend/app/api/building.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.base import get_db
from app.models.building import Building
from app.schemas.building import BuildingCreate, BuildingResponse, BuildingUpdate
from app.services.building_service import BuildingService

router = APIRouter(tags=["Buildings"])


@router.post("/buildings", response_model=BuildingResponse, status_code=201)
async def create_building(data: BuildingCreate, db: AsyncSession = Depends(get_db)) -> Building:
    service = BuildingService(db)
    return await service.create_building(data)


@router.get("/buildings", response_model=list[BuildingResponse])
async def list_buildings(db: AsyncSession = Depends(get_db)) -> list[Building]:
    service = BuildingService(db)
    return await service.get_all_buildings()


@router.get("/buildings/{building_id}", response_model=BuildingResponse)
async def get_building(building_id: int, db: AsyncSession = Depends(get_db)) -> Building:
    service = BuildingService(db)
    return await service.get_building(building_id)


@router.put("/buildings/{building_id}", response_model=BuildingResponse)
async def update_building(
    building_id: int, data: BuildingUpdate, db: AsyncSession = Depends(get_db)
) -> Building:
    service = BuildingService(db)
    return await service.update_building(building_id, data)


@router.delete("/buildings/{building_id}", status_code=204)
async def delete_building(building_id: int, db: AsyncSession = Depends(get_db)) -> None:
    service = BuildingService(db)
    await service.delete_building(building_id)

File: frontend/src/services/building.ts
import { api } from "@/services/api";
import type { Building, BuildingCreateInput, BuildingUpdateInput } from "@/types/building";

export const buildingService = {
  list: async (): Promise<Building[]> => {
    const { data } = await api.get<Building[]>("/buildings");
    return data;
  },

  get: async (id: number): Promise<Building> => {
    const { data } = await api.get<Building>(`/buildings/${id}`);
    return data;
  },

  create: async (payload: BuildingCreateInput): Promise<Building> => {
    const { data } = await api.post<Building>("/buildings", payload);
    return data;
  },

  update: async (id: number, payload: BuildingUpdateInput): Promise<Building> => {
    const { data } = await api.put<Building>(`/buildings/${id}`, payload);
    return data;
  },

  remove: async (id: number): Promise<void> => {
    await api.delete(`/buildings/${id}`);
  },
};
